libs/libcommon/src/libcommon/utils.py

```diff
@@ -148,3 +155,68 @@ def raise_if_blocked(
                 "This dataset has been disabled for now. Please open an issue in"
                 " https://github.com/huggingface/datasets-server if you want this dataset to be supported."
             )
+
+
+FuncT = TypeVar("FuncT", bound=Callable[..., Any])
+RETRY_SLEEPS = (1, 1, 1, 10, 10, 10, 60, 60, 60, 10 * 60)
+RETRY_ON: tuple[type[Exception]] = (Exception,)
+
+
+class retry:
+    """retries with an increasing sleep before every attempt"""
+
+    def __init__(self, sleeps: Sequence[float] = RETRY_SLEEPS, on: Sequence[type[Exception]] = RETRY_ON) -> None:
+        self.sleeps = sleeps
+        self.on = on
+
+    def __call__(self, func: FuncT) -> FuncT:
+        @functools.wraps(func)
+        def decorator(*args: Any, **kwargs: Any) -> Any:
+            attempt = 0
+            last_err = None
+            while attempt < len(self.sleeps):
+                try:
+                    """always sleep before calling the function. It will prevent rate limiting in the first place"""
+                    duration = self.sleeps[attempt]
+                    logging.info(f"Sleep during {duration} seconds to preventively mitigate rate limiting.")
+                    time.sleep(duration)
+                    return func(*args, **kwargs)
+                except tuple(self.on) as err:
+                    logging.info(f"Got a {type(err)}. Let's retry.")
+                    last_err = err
+                    attempt += 1
+            raise RuntimeError(f"Give up after {attempt} attempts. The last one raised {type(last_err)}") from last_err
+
+        return cast(FuncT, decorator)
+
+
+HF_HUB_HTTP_ERROR_RETRY_SLEEPS = [1, 1, 1, 10, 10, 10]
+
+
+def download_file_from_hub(
+    repo_type: str,
+    revision: str,
+    repo_id: str,
+    filename: str,
+    local_dir: Union[str, Path],
+    hf_token: Optional[str],
+    cache_dir: Union[str, Path, None] = None,
+    force_download: bool = False,
+    resume_download: bool = False,
+) -> None:
+    # Force hf_transfer usage
+    constants.HF_HUB_ENABLE_HF_TRANSFER = True
+    logging.debug(f"Using {constants.HF_HUB_ENABLE_HF_TRANSFER} for hf_transfer")
+    retry_download_hub_file = retry(on=[ReadTimeout], sleeps=HF_HUB_HTTP_ERROR_RETRY_SLEEPS)(hf_hub_download)
```

Comment:
When using `hf_transfer` you don't have clear Python exception since it's a raw rust-based package. [You will receive a `RuntimeError` instead](https://github.com/huggingface/huggingface_hub/blob/969af395e3137ea508c6db01f27dd9a111fc6050/src/huggingface_hub/file_download.py#L516).

Also FYI, there is a built-in mechanism in `hf_hub_download` to retry on `(requests.ConnectionError, requests.ReadTimeout)` up to five times with 1s delay in between. This means that in case of error, there will be 10 (from datasets-server) x 5 (from hfh) retries for the same file. Not a problem in itself, just a "good to know" :)  (and this is only when hf_transfer is NOT used)